src/Utilities/DoNotCatchGeneralUnlessRethrown.cs

```diff
@@ -0,0 +1,149 @@
+        {
+            _shouldCheckLambdas = shouldCheckLambdas;
+            _enablingMethodAttributeFullyQualifiedName = enablingMethodAttributeFullyQualifiedName;
+        }
+
+        protected abstract Diagnostic CreateDiagnostic(IMethodSymbol containingMethod, SyntaxNode catchNode);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            analysisContext.RegisterCompilationStartAction(compilationStartAnalysisContext =>
+            {
+                INamedTypeSymbol requiredAttributeType = null;
+                if (RequiresAttributeOnMethod && (requiredAttributeType = GetRequiredAttributeType(compilationStartAnalysisContext.Compilation)) == null)
+                {
+                    return;
+                }
+
+                var disallowedCatchTypes = GetDisallowedCatchTypes(compilationStartAnalysisContext.Compilation);
+
+                compilationStartAnalysisContext.RegisterOperationBlockAction(operationBlockAnalysisContext =>
+                {
+                    if (operationBlockAnalysisContext.OwningSymbol.Kind != SymbolKind.Method)
+                    {
+                        return;
+                    }
+
+                    var method = (IMethodSymbol)operationBlockAnalysisContext.OwningSymbol;
+
+                    if (RequiresAttributeOnMethod && !MethodHasAttribute(method, requiredAttributeType))
+                    {
+                        return;
+                    }
+
+                    foreach (var operation in operationBlockAnalysisContext.OperationBlocks)
+                    {
+                        var walker = new DisallowGeneralCatchUnlessRethrowWalker(disallowedCatchTypes, _shouldCheckLambdas);
+                        walker.Visit(operation);
+
+                        foreach (var catchClause in walker.CatchClausesForDisallowedTypesWithoutRethrow)
+                        {
+                            operationBlockAnalysisContext.ReportDiagnostic(CreateDiagnostic(method, catchClause.Syntax));
+                        }
+                    }
+                });
+            });
+        }
+
+        private INamedTypeSymbol GetRequiredAttributeType(Compilation compilation)
+        {
+            return compilation.GetTypeByMetadataName(_enablingMethodAttributeFullyQualifiedName);
+        }
+
+        private bool MethodHasAttribute(IMethodSymbol method, INamedTypeSymbol attributeType)
+        {
+            return method.GetAttributes().Any(attribute => attribute.AttributeClass.Equals(attributeType));
+        }
+
+        private static ICollection<INamedTypeSymbol> GetDisallowedCatchTypes(Compilation compilation)
+        {
+            var disallowed = new List<INamedTypeSymbol>();
```

Comment:
>new List [](start = 29, length = 8)

Can we use `ImmutableHashSet.Builder` here and return an `ImmutableHashSet` instead? Also please ensure you filter out null well known types (consider adding an extension method AddIfNotNull if suitable).

Reply:
Done